src/main/kotlin/uk/gov/gdx/datashare/services/ConsumersService.kt

```diff
@@ -14,45 +15,96 @@ import java.util.*
+    }
+  }
 
-  fun getSubscriptionsForConsumer(consumerId: UUID) =
-    consumerSubscriptionRepository.findAllByConsumerId(consumerId)
+  fun mapConsumerSubscriptionDto(
+    consumerSubscription: ConsumerSubscription,
+    enrichmentFields: List<ConsumerSubscriptionEnrichmentField>,
+  ): ConsumerSubscriptionDto {
+    return ConsumerSubscriptionDto(
+      consumerSubscriptionId = consumerSubscription.consumerSubscriptionId,
+      consumerId = consumerSubscription.consumerId,
+      oauthClientId = consumerSubscription.oauthClientId,
+      eventType = consumerSubscription.eventType,
+      enrichmentFields = enrichmentFields.map { it.enrichmentField },
+      enrichmentFieldsIncludedInPoll = consumerSubscription.enrichmentFieldsIncludedInPoll,
+      whenCreated = consumerSubscription.whenCreated,
+    )
+  }
+
+  fun addConsumerSubscriptionEnrichmentFields(
+    consumerSubscriptionId: UUID,
+    enrichmentFields: List<String>,
+  ): List<ConsumerSubscriptionEnrichmentField> {
+    return consumerSubscriptionEnrichmentFieldRepository.saveAll(
+      enrichmentFields.map {
+        ConsumerSubscriptionEnrichmentField(
+          consumerSubscriptionId = consumerSubscriptionId,
+          enrichmentField = it,
+        )
+      },
+    ).toList()
+  }
 
   fun addConsumerSubscription(
     consumerId: UUID,
     consumerSubRequest: ConsumerSubRequest,
-  ): ConsumerSubscription {
+  ): ConsumerSubscriptionDto {
     with(consumerSubRequest) {
-      return consumerSubscriptionRepository.save(
+      val consumerSubscription = consumerSubscriptionRepository.save(
         ConsumerSubscription(
           consumerId = consumerId,
           oauthClientId = oauthClientId,
           eventType = eventType,
-          enrichmentFields = enrichmentFields,
         ),
       )
+      val enrichmentFields =
+        addConsumerSubscriptionEnrichmentFields(consumerSubscription.consumerSubscriptionId, enrichmentFields)
+
+      return mapConsumerSubscriptionDto(consumerSubscription, enrichmentFields)
     }
   }
 
   fun updateConsumerSubscription(
     consumerId: UUID,
     subscriptionId: UUID,
     consumerSubRequest: ConsumerSubRequest,
-  ): ConsumerSubscription {
+  ): ConsumerSubscriptionDto {
     with(consumerSubRequest) {
-      return consumerSubscriptionRepository.save(
+      val consumerSubscription = consumerSubscriptionRepository.save(
         consumerSubscriptionRepository.findByIdOrNull(subscriptionId)?.copy(
           consumerId = consumerId,
           oauthClientId = oauthClientId,
           eventType = eventType,
-          enrichmentFields = enrichmentFields,
           enrichmentFieldsIncludedInPoll = enrichmentFieldsIncludedInPoll,
         ) ?: throw ConsumerSubscriptionNotFoundException("Subscription $subscriptionId not found"),
       )
+
+      consumerSubscriptionEnrichmentFieldRepository.deleteAllByConsumerSubscriptionId(consumerSubscription.consumerSubscriptionId)
```

Comment:
Instead of deleting all of these, could we delete only the extra and update with the new ones, otherwise its a bit too much. We could also do all this update in a single query if you would rather write SQL

Reply:
I chatted to olly about this, and he reckoned it would be better to delete and add them all, as the logic is a bit cleaner and it's still just two calls

Reply:
Sure, SGTM :)